src/Shared/Microsoft.DotNet.Internal.Health/ServiceFabricHealthReportProvider.cs

```diff
@@ -0,0 +1,115 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Fabric;
+using System.Fabric.Health;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.DotNet.Internal.Health
+{
+    public class ServiceFabricHealthReportOptions
+    {
+        public string ServiceUri { get; set; }
+    }
+
+    public class ServiceFabricHealthReportProvider : IHealthReportProvider
+    {
+        private readonly IOptions<ServiceFabricHealthReportOptions> _options;
+        private readonly ServiceContext _context;
+        private readonly FabricClient _fabricClient;
+        private readonly bool _isStateful;
+
+        public ServiceFabricHealthReportProvider(StatefulServiceContext statefulServiceContext, IOptions<ServiceFabricHealthReportOptions> options) : this(statefulServiceContext, options, true)
+        {
+        }
+
+        public ServiceFabricHealthReportProvider(StatelessServiceContext statelessServiceContext, IOptions<ServiceFabricHealthReportOptions> options) : this(statelessServiceContext, options, false)
+        {
+        }
+
+        private ServiceFabricHealthReportProvider(ServiceContext context, IOptions<ServiceFabricHealthReportOptions> options, bool isStateful)
+        {
+            _context = context;
+            _options = options;
+            _isStateful = isStateful;
+            _fabricClient = new FabricClient();
+        }
+
+        private Uri GetServiceUri()
+        {
+            string optionsUri = _options.Value.ServiceUri;
+            if (!string.IsNullOrEmpty(optionsUri))
+                return new Uri(optionsUri);
+
+            return _context.ServiceName;
+        }
+
+        public Task UpdateStatusAsync(string serviceName, string instance, string subStatusName, HealthStatus status, string message)
+        {
+            var healthInfo = new HealthInformation(
+                GetType().FullName,
+                subStatusName,
+                MapStatus(status))
+            {
+                Description = message
+            };
+
+            System.Fabric.Health.HealthReport report;
+            if (instance == null)
+            {
+                report = new ServiceHealthReport(
+                    GetServiceUri(),
+                    healthInfo
+                );
+            }
+            else if (_isStateful)
+            {
+                report = new StatefulServiceReplicaHealthReport(
+                    _context.PartitionId,
```

Comment:
I think partitionId isn't the right id here. I think this needs to be the ReplicaOrInstanceId property

Reply:
The parameter is called "partitionId", and it's a Guid. The second parameter is the ReplicaOrInstanceId. I'm not 100% sure why you need to specifiy both... but there you have it.

Reply:
Hrm... the fact that ServiceFabric requires all this identity stuff in weird formats is making this really hard to generalize.  Darn SF...

Reply:
I stopped using the input "instance", and just used the one on _context. I think that's the best compromise.